filter system paths

davydden opened the pull request:
fixes https://github.com/LLNL/spack/issues/2079
fixes https://github.com/LLNL/spack/issues/1224


~~If this is a right way, i will look through `cc` to filter it as well.~~

tgamblin commented on lib/spack/spack/util/environment.py:
```diff
@@ -25,6 +25,27 @@
+    Arguments:
+    paths -- array of paths
+    """
+    system_paths = ['/usr']
```
You probably want to be a bit more specific:
  1. We care about `lib`, `lib64`, and `include` in system locations, and we want them stripped from `-I`, `-L`, and `-Wl,-rpath`.
  2. We also care about `bin` for `PATH`, but I think we still need system paths to be added to `PATH` at lower precedence than any `PATH` from a Spack package.

For (1), consider what happens when `Homebrew` is installed in /usr/local/Homebrew and you try to depend on a homebrew keg-only package.  It'll be stripped by these rules, but it's actually not in any system path.

So why not:

```python
system_paths = ['/', '/usr/', '/usr/local']
suffixes = ['lib', 'lib64', 'include']
system_dirs = sum(([os.path.join(p, s)
                    for s in suffixes] for p in system_paths), [])

def filter_system_paths(paths):
    return [p for p in paths if not any(p in s for s in system_dirs)]
```

which gives:
```
>>> filter_system_paths([
    '/usr/local/Cellar/gcc/5.3.0/lib',
    '/usr/local/lib',
    '/usr/local/include',
    '/usr/local/lib64',
    '/usr/local/opt/some-package/lib',
    '/usr/opt/lib',
    '/lib',
    '/include',
    '/opt/some-package/include',
])
['/usr/local/Cellar/gcc/5.3.0/lib', '/usr/local/opt/some-package/lib', '/usr/opt/lib', '/opt/some-package/include']
```

We can also later extract `system_paths` and stick it in a `config.yaml`.

davydden commented:
@tgamblin I updated the filtering functions and added another one for `bin`. Looking at `cc`, it looks like nothing is to be done there as it uses SPACK environment variables.

davydden commented:
`$ spack install arpack-ng~mpi ^netlib-lapack` with _system_ blas/lapack gives the following relevant environment:

```
CMAKE_PREFIX_PATH=/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/cmake-3.6.1-knsjcg5wlbnsy27ja5ngp5xdftflnv4a:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/bzip2-1.0.6-txhecghwl6dhscbg62hiejwoqfvhasoo:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/curl-7.50.3-ktu7vhoi6p4fmbldeapjqzvkvct6z6dk:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/openssl-1.0.2j-k6g3fa3rrglj5wenvowtaruuvb6nmn6t:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/zlib-1.2.8-yefo4t2eiiwps7wrbvvgauo2z4ozxlyg:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/expat-2.2.0-lmk3zwfdwzemdjrrkrehldzjpuvmkbxe:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libarchive-3.2.1-a6vndqjqcae2rtj7tk4qcvqoaiin3nfr:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libxml2-2.9.4-id5me6ic2r5eqcozfuhfg2pbyx4txgd6:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/xz-5.2.2-l3ntfowzw5tk2zbs3bmuy2nei4i3rqzc:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lz4-131-e6cru3imydxci7aaskrssjo6vjya55qy:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lzma-4.32.7-2ka7vkselplzy5hj3iyvquej7frhidjk:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lzo-2.09-2dk4mx6m2kg65y34ndw6t3eh2iq7glkx:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/nettle-3.2-fioxw3hgk3gp57os6sxqfvmmlijad6al:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/gmp-6.1.1-vj3e2ygpbbbo2mmyomdzuizeboramub7:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/m4-1.4.17-r5nzualkj7pwaflxtt2mxifhxfkefmpn:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libsigsegv-2.10-cotbexnf4mm6u2eueit5uorf3xtvxbpw:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/ncurses-6.0-sflzz4eoc2zxhxje5z3rtcbylrujuu4h
PATH=/usr/bin:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/cmake-3.6.1-knsjcg5wlbnsy27ja5ngp5xdftflnv4a/bin:/Users/davydden/spack/lib/spack/env:/Users/davydden/spack/lib/spack/env/case-insensitive:/Users/davydden/spack/lib/spack/env/clang:/Users/davydden/spack/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Library/TeX/texbin
PKG_CONFIG_PATH=/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/nettle-3.2-fioxw3hgk3gp57os6sxqfvmmlijad6al/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lz4-131-e6cru3imydxci7aaskrssjo6vjya55qy/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/xz-5.2.2-l3ntfowzw5tk2zbs3bmuy2nei4i3rqzc/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libxml2-2.9.4-id5me6ic2r5eqcozfuhfg2pbyx4txgd6/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libarchive-3.2.1-a6vndqjqcae2rtj7tk4qcvqoaiin3nfr/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/expat-2.2.0-lmk3zwfdwzemdjrrkrehldzjpuvmkbxe/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/zlib-1.2.8-yefo4t2eiiwps7wrbvvgauo2z4ozxlyg/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/openssl-1.0.2j-k6g3fa3rrglj5wenvowtaruuvb6nmn6t/lib/pkgconfig:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/curl-7.50.3-ktu7vhoi6p4fmbldeapjqzvkvct6z6dk/lib/pkgconfig
SPACK_DEPENDENCIES=/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/cmake-3.6.1-knsjcg5wlbnsy27ja5ngp5xdftflnv4a:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/bzip2-1.0.6-txhecghwl6dhscbg62hiejwoqfvhasoo:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/curl-7.50.3-ktu7vhoi6p4fmbldeapjqzvkvct6z6dk:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/openssl-1.0.2j-k6g3fa3rrglj5wenvowtaruuvb6nmn6t:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/zlib-1.2.8-yefo4t2eiiwps7wrbvvgauo2z4ozxlyg:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/expat-2.2.0-lmk3zwfdwzemdjrrkrehldzjpuvmkbxe:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libarchive-3.2.1-a6vndqjqcae2rtj7tk4qcvqoaiin3nfr:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libxml2-2.9.4-id5me6ic2r5eqcozfuhfg2pbyx4txgd6:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/xz-5.2.2-l3ntfowzw5tk2zbs3bmuy2nei4i3rqzc:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lz4-131-e6cru3imydxci7aaskrssjo6vjya55qy:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lzma-4.32.7-2ka7vkselplzy5hj3iyvquej7frhidjk:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/lzo-2.09-2dk4mx6m2kg65y34ndw6t3eh2iq7glkx:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/nettle-3.2-fioxw3hgk3gp57os6sxqfvmmlijad6al:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/gmp-6.1.1-vj3e2ygpbbbo2mmyomdzuizeboramub7:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/m4-1.4.17-r5nzualkj7pwaflxtt2mxifhxfkefmpn:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/libsigsegv-2.10-cotbexnf4mm6u2eueit5uorf3xtvxbpw:/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/ncurses-6.0-sflzz4eoc2zxhxje5z3rtcbylrujuu4h
SPACK_ENV_PATH=/Users/davydden/spack/lib/spack/env:/Users/davydden/spack/lib/spack/env/case-insensitive:/Users/davydden/spack/lib/spack/env/clang
SPACK_LINK_DEPS=
SPACK_RPATH_DEPS=
```

`/usr` is only present in `PATH` (as it should), but strangely `PATH` starts from `/usr/bin`...

davydden commented:
> but strangely PATH starts from /usr/bin...

that's because `prepend_path` for `PATH`:

```
bin_dirs = reversed(filter(os.path.isdir, [
        '%s/bin' % d.prefix for d in pkg.spec.dependencies(deptype='build')]))
bin_dirs = filter_system_bin_paths(bin_dirs)
for item in bin_dirs:
        env.prepend_path('PATH', item)
```

I will adjust the function...

davydden commented:
now the path is better:
```
PATH=/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/cmake-3.6.1-knsjcg5wlbnsy27ja5ngp5xdftflnv4a/bin:/usr/bin:/Users/davydden/spack/lib/spack/env:/Users/davydden/spack/lib/spack/env/case-insensitive:/Users/davydden/spack/lib/spack/env/clang:/Users/davydden/spack/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin:/Library/TeX/texbin
```

davydden commented:
@adamjstewart @alalazo any comments from your side for this filtering of paths?

alalazo commented on lib/spack/spack/util/environment.py:
```diff
@@ -24,6 +24,32 @@
+system_paths = ['/', '/usr/', '/usr/local']
+suffixes = ['lib', 'lib64', 'include']
+system_dirs = sum(([os.path.join(p, s)
+                    for s in suffixes] for p in system_paths), [])
```
Here you can use a single list comprehension : 
```python
system_dirs = [os.path.join(p, s) for s in suffixes for p in system_paths]
```

alalazo commented on lib/spack/spack/util/environment.py:
```diff
@@ -24,6 +24,32 @@
+
+
+def filter_system_paths(paths):
+    return [p for p in paths if not any(p in s for s in system_dirs)]
```
It seems to me that you are trying exact match here. If I am not wrong then:
```python
[p for p in paths if p not in system_dirs]
```
should do the same thing.

alalazo commented on lib/spack/spack/util/environment.py:
```diff
@@ -24,6 +24,32 @@
+            filtered.append(p)
+
+    # add to the begining because PATH is constructed with prepend_path(item)
+    filtered[:0] = bins
```
What about the following :
```python
def filter_system_bin_paths(paths):
    bins = [p for p in paths if p in system_bins]
    filtered = dedupe(bins + paths)
    return filtered
```
Would this work ?

davydden replied:
otherwise i think your code should work and is certainly more elegant :smile:

alalazo replied:
Sorry, I missed that : 
```
from llnl.util.lang import dedupe
```

davydden replied:
the result is `<generator object dedupe at 0x102c1ac30>`... something is strange...

davydden replied:
ps. happens in the unit test:
```
AssertionError: <generator object dedupe at 0x102c1ac30> != ['/usr/local/bin', '/bin', '/usr/local/Cellar/gcc/5.3.0/bin', '/usr/local/opt/some-package/bin', '/usr/opt/bin', '/opt/some-package/bin']
```

alalazo replied:
My fault : 
```python
filtered = [x for x in dedupe(bins + paths)]
```
I made it a generator to avoid constructing a list if the temporary was big.

alalazo replied:
+1 for your last change

davydden commented:
@alalazo thanks for reviewing and improvements suggestions.

davydden commented:
somehow tests failed (or something failed) on `travis`, so I try another alternative to filter_bins without `dedupe`.

davydden commented:
Travis errors seem to be unrelated

alalazo commented:
@davydden I'll have a look

alalazo commented:
@davydden I opened a pull request to your branch. You can test if this works for you doing : 
```
spack test
```
locally. If it works feel free to either merge it or adapt it manually to another implementation you like better.

davydden commented:
@alalazo thanks, i am trying something along the lines of your PR...

davydden commented:
@tgamblin :   the tests pass on this PR, should be ready.

tgamblin commented:
@davydden: this LGTM.  Only thing I worry about is whether it will break some packages that depend on `libtool` on OS X, as it's preferred to use the system libtool there.  I think we can work that out separately.  Thoughts?

davydden commented:
@tgamblin :  The only package I am aware of in Spack which has extra hacks for libtool on macOS is [boost](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/boost/package.py#L262-L270), this PR should not affect this anyhow. 

Here we only filter out thinks which come from external packages, so all default behaviour should stay ok.

tgamblin commented:
@davydden @alalazo @adamjstewart @citibeth: now that this is merged, we can take out the sections that mention false paths in the docs:
* http://spack.readthedocs.io/en/latest/configuration.html#false-paths-for-system-packages
* http://spack.readthedocs.io/en/latest/getting_started.html#openssl
Does someone want to take a stab at that?

citibeth commented:
Done

On Wed, Oct 26, 2016 at 7:34 PM, Todd Gamblin <notifications@github.com>
wrote:

> @davydden <https://github.com/davydden> @alalazo
> <https://github.com/alalazo> @adamjstewart
> <https://github.com/adamjstewart> @citibeth <https://github.com/citibeth>:
> now that this is merged, we can take out the sections that mention false
> paths in the docs:
>
>    - http://spack.readthedocs.io/en/latest/configuration.html#
>    false-paths-for-system-packages
>    <http://spack.readthedocs.io/en/latest/configuration.html#false-paths-for-system-packages>
>    - http://spack.readthedocs.io/en/latest/getting_started.html#openssl
>    Does someone want to take a stab at that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/2083#issuecomment-256507244>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd0Bg0rbzekIIfrTt0E6MqCRGMFD5ks5q3-N3gaJpZM4Kd9ak>
> .
>

tgamblin commented:
Thanks!
